### PR TITLE
feat: support horizontal rules on matrix

### PR DESCRIPTION
**Pull request by sehilyi**
Allows using vertical rules overlaid on top of a matrix (see below example). This PR also fixes an issue that the x and y axes were not correctly shown in matrix visualization (e.g., left axis being shown on the right or top axis being hidden).

Follow up on #615

```js
{
    alignment: 'overlay',
    tracks: [
        {
            data: {
                url: GOSLING_PUBLIC_DATA.matrix,
                type: 'matrix'
            },
            mark: 'bar',
            x: { field: 'xs', type: 'genomic', axis: 'none' },
            xe: { field: 'xe', type: 'genomic', axis: 'none' },
            y: { field: 'ys', type: 'genomic', axis: 'none' },
            ye: { field: 'ye', type: 'genomic', axis: 'none' },
            color: { field: 'value', type: 'quantitative', range: 'grey', legend: true },
            style: { background: 'lightgray' }
        },
        {
            data: {
                type: 'json',
                values: [
                    { c: 'chr2', p: 100000 },
                    { c: 'chr5', p: 100000 },
                    { c: 'chr10', p: 100000 }
                ],
                chromosomeField: 'c',
                genomicFields: ['p']
            },
            mark: 'rule',
            x: { field: 'p', type: 'genomic', axis: 'none' },
            strokeWidth: { value: 2 },
            color: { value: 'red' }
        },
        {
            data: {
                type: 'json',
                values: [
                    { c: 'chr2', p: 100000 },
                    { c: 'chr5', p: 100000 },
                    { c: 'chr10', p: 100000 }
                ],
                chromosomeField: 'c',
                genomicFields: ['p']
            },
            mark: 'rule',
            y: { field: 'p', type: 'genomic', axis: 'none' },
            strokeWidth: { value: 2 },
            color: { value: 'blue' }
        }
    ],
    width: 600,
    height: 600,
    style: { dashed: [3, 3] }
}
```
gosling-visualization (2)

### CODE REVIEW

**Comment by sehilyi** [comment]
Screen Shot 2022-01-21 at 14 22 31

- [ ] Legend should be positioned on top of the rules (https://github.com/gosling-lang/gosling.js/issues/473)

Fix this in a separate PR

File: editor/example/index.ts [hunk 1]
```diff
@@ -62,7 +62,8 @@ export const examples: {
     MATRIX: {
         name: 'Basic Example: Hi-C Matrix',
         spec: EX_SPEC_MATRIX,
-        underDevelopment: true
+        underDevelopment: true,
+        forceShow: true
```
Comment (sehilyi) [inline editor/example/index.ts]: Should reset before merge

File: editor/example/matrix.ts [hunk 2]
```diff
@@ -9,30 +9,30 @@ export const EX_SPEC_MATRIX: GoslingSpec = {
...
-                                axis: 'top'
-                            },
-                            y: { field: 'peak', type: 'quantitative' },
-                            color: { field: 'sample', type: 'nominal' },
-                            width: 600,
-                            height: 50
-                        }
-                    ]
-                },
+                // {
```
Comment (sehilyi) [inline editor/example/matrix.ts]: Uncomment this part after allowing axes

File: src/core/mark/index.ts [hunk 3]
```diff
@@ -69,7 +69,7 @@ export function drawMark(HGC: any, trackInfo: any, tile: any, model: GoslingTrac
         // }
     });
 
-    if (Is2DTrack(model.spec())) {
+    if (Is2DTrack(model.spec()) || IsVerticalRule(model.spec())) {
```
Comment (sehilyi) [inline src/core/mark/index.ts:72]: We could simply check whether y-axis is mapped with a `genomic` field

File: src/core/gosling-to-higlass.ts [hunk 4]
```diff
@@ -210,7 +210,7 @@ export function goslingToHiGlass(
             ) {
                 const narrowType = getAxisNarrowType(c as any, gosTrack.orientation, bb.width, bb.height);
                 hgModel.setAxisTrack(channel.axis, narrowType, {
-                    id: `${firstResolvedSpec.id}-axis`,
+                    id: `${firstResolvedSpec.id}-${channel.axis}-axis`,
```
Comment (sehilyi) [inline src/core/gosling-to-higlass.ts:213]: Added to use different IDs for both x and y axes. If the same ID is being used, HiGlass only renders one of them.

File: src/core/higlass-model.ts [hunk 5]
```diff
@@ -338,11 +338,11 @@ export class HiGlassModel {
                 }
                 this.getLastView().tracks.left = insertItemToArray(this.getLastView().tracks.left, 0, axisTrack);
             } else if (position === 'right') {
-                if (this.getLastView().tracks.left.filter(d => d.type === 'axis-track').length !== 0) {
+                if (this.getLastView().tracks.right.filter(d => d.type === 'axis-track').length !== 0) {
```
Comment (sehilyi) [inline src/core/higlass-model.ts:341]: There was a typo that 'right' was written as 'left'

File: src/core/utils/bounding-box.ts [hunk 6]
```diff
@@ -213,7 +214,7 @@ function traverseAndCollectTrackInfo(
                 }
             });
         } else {
-            // This is a horizontal view, so use the largest `width` for this view.
+            // This is a horizontal, so use the largest `width` for this view.
```
Comment (sehilyi) [inline src/core/utils/bounding-box.ts]:
```suggestion
            // This is a horizontal view, so use the largest `width` for this view.
```

**Comment by sehilyi** [comment]
Hi @manzt, just want to share that this is a follow-up PR of #615 that enables horizontal rules on a matrix visualization, so when this is merged, both the horizontal and vertical rules can be overlaid on top of a matrix. There are some relatively minor issues to address, e.g., adjusting the z-index. Feel free to review this if time permits for you.

**Comment by manzt** [comment]
Thanks @sehilyi, this looks great! Sorry if you were looking for a review and I missed it. Feel free to request a review whenever if the future if you'd like. (It's easier for me to keep track of PR review requests if assigned. Sometimes I miss this if it's in the comment).

Can't wait to try out this new functionality.

**Comment by sehilyi** [comment]
@manzt Sure, I will assign you as a reviewer next time. Thanks!

**Pull request merged**